diff --git a/internals/hooks/useField/useFieldV6TextField.js b/internals/hooks/useField/useFieldV6TextField.js
index 8c4f20ac95fb7385f03faffeefdcda39bb7dcd81..efddab05aa7404bb8181225c9828051143c46987 100644
--- a/internals/hooks/useField/useFieldV6TextField.js
+++ b/internals/hooks/useField/useFieldV6TextField.js
@@ -88,7 +88,7 @@ export const useFieldV6TextField = params => {
       // On multi input range pickers we want to update selection range only for the active input
       // This helps to avoid the focus jumping on Safari https://github.com/mui/mui-x/issues/9003
       // because WebKit implements the `setSelectionRange` based on the spec: https://bugs.webkit.org/show_bug.cgi?id=224425
-      if (inputRef.current !== getActiveElement(document)) {
+      if (inputRef.current !== getActiveElement(inputRef.current?.ownerDocument)) {
         return;
       }
 
@@ -101,7 +101,7 @@ export const useFieldV6TextField = params => {
         const selectionStart = selectedSection.type === 'empty' ? selectedSection.startInInput - selectedSection.startSeparator.length : selectedSection.startInInput;
         const selectionEnd = selectedSection.type === 'empty' ? selectedSection.endInInput + selectedSection.endSeparator.length : selectedSection.endInInput;
         if (selectionStart !== inputRef.current.selectionStart || selectionEnd !== inputRef.current.selectionEnd) {
-          if (inputRef.current === getActiveElement(document)) {
+          if (inputRef.current === getActiveElement(inputRef.current?.ownerDocument)) {
             inputRef.current.setSelectionRange(selectionStart, selectionEnd);
           }
         }
@@ -109,7 +109,7 @@ export const useFieldV6TextField = params => {
         selectionSyncTimeoutRef.current = setTimeout(() => {
           // handle case when the selection is not updated correctly
           // could happen on Android
-          if (inputRef.current && inputRef.current === getActiveElement(document) &&
+          if (inputRef.current && inputRef.current === getActiveElement(inputRef.current?.ownerDocument) &&
           // The section might loose all selection, where `selectionStart === selectionEnd`
           // https://github.com/mui/mui-x/pull/13652
           inputRef.current.selectionStart === inputRef.current.selectionEnd && (inputRef.current.selectionStart !== selectionStart || inputRef.current.selectionEnd !== selectionEnd)) {
@@ -132,14 +132,14 @@ export const useFieldV6TextField = params => {
       return nextSectionIndex === -1 ? sections.length - 1 : nextSectionIndex - 1;
     },
     focusField: (newSelectedSection = 0) => {
-      if (getActiveElement(document) === inputRef.current) {
+      if (getActiveElement(inputRef.current?.ownerDocument) === inputRef.current) {
         return;
       }
       inputRef.current?.focus();
       setSelectedSections(newSelectedSection);
     },
     setSelectedSections: newSelectedSections => setSelectedSections(newSelectedSections),
-    isFieldFocused: () => inputRef.current === getActiveElement(document)
+    isFieldFocused: () => inputRef.current === getActiveElement(inputRef.current?.ownerDocument)
   }), [inputRef, parsedSelectedSections, sections, setSelectedSections]);
   const syncSelectionFromDOM = () => {
     const browserStartIndex = inputRef.current.selectionStart ?? 0;
@@ -298,7 +298,7 @@ export const useFieldV6TextField = params => {
   const valueStr = React.useMemo(() => state.tempValueStrAndroid ?? fieldValueManager.getV6InputValueFromSections(state.sections, localizedDigits, isRtl), [state.sections, fieldValueManager, state.tempValueStrAndroid, localizedDigits, isRtl]);
   React.useEffect(() => {
     // Select all the sections when focused on mount (`autoFocus = true` on the input)
-    if (inputRef.current && inputRef.current === getActiveElement(document)) {
+    if (inputRef.current && inputRef.current === getActiveElement(inputRef.current?.ownerDocument)) {
       setSelectedSections('all');
     }
     return () => {
@@ -316,7 +316,7 @@ export const useFieldV6TextField = params => {
     }
     return 'numeric';
   }, [activeSectionIndex, state.sections]);
-  const inputHasFocus = inputRef.current && inputRef.current === getActiveElement(document);
+  const inputHasFocus = inputRef.current && inputRef.current === getActiveElement(inputRef.current?.ownerDocument);
   const shouldShowPlaceholder = !inputHasFocus && areAllSectionsEmpty;
   return {
     interactions,
diff --git a/internals/hooks/useField/useFieldV7TextField.js b/internals/hooks/useField/useFieldV7TextField.js
index cb5e3d2ee510df8e878b849cdd513563af7337b9..82ce49ac90bb50fb59f76ccac098491d8de9ba0e 100644
--- a/internals/hooks/useField/useFieldV7TextField.js
+++ b/internals/hooks/useField/useFieldV7TextField.js
@@ -48,7 +48,7 @@ export const useFieldV7TextField = params => {
       if (!sectionListRef.current) {
         return;
       }
-      const selection = document.getSelection();
+      const selection = sectionListRef.current?.getRoot?.()?.ownerDocument.getSelection();
       if (!selection) {
         return;
       }
@@ -64,7 +64,7 @@ export const useFieldV7TextField = params => {
       }
 
       // On multi input range pickers we want to update selection range only for the active input
-      if (!sectionListRef.current.getRoot().contains(getActiveElement(document))) {
+      if (!sectionListRef.current.getRoot().contains(getActiveElement(sectionListRef.current?.getRoot?.()?.ownerDocument))) {
         return;
       }
       const range = new window.Range();
@@ -85,7 +85,7 @@ export const useFieldV7TextField = params => {
       selection.addRange(range);
     },
     getActiveSectionIndexFromDOM: () => {
-      const activeElement = getActiveElement(document);
+      const activeElement = getActiveElement(sectionListRef.current?.getRoot?.()?.ownerDocument);
       if (!activeElement || !sectionListRef.current || !sectionListRef.current.getRoot().contains(activeElement)) {
         return null;
       }
@@ -111,7 +111,7 @@ export const useFieldV7TextField = params => {
       setSelectedSections(newSelectedSections);
     },
     isFieldFocused: () => {
-      const activeElement = getActiveElement(document);
+      const activeElement = getActiveElement(sectionListRef.current?.getRoot?.()?.ownerDocument);
       return !!sectionListRef.current && sectionListRef.current.getRoot().contains(activeElement);
     }
   }), [parsedSelectedSections, setSelectedSections, state.sections, focused]);
@@ -138,7 +138,7 @@ export const useFieldV7TextField = params => {
     onClick?.(event, ...args);
     if (parsedSelectedSections === 'all') {
       setTimeout(() => {
-        const cursorPosition = document.getSelection().getRangeAt(0).startOffset;
+        const cursorPosition = sectionListRef.current?.getRoot?.()?.ownerDocument.getSelection().getRangeAt(0).startOffset;
         if (cursorPosition === 0) {
           setSelectedSections(sectionOrder.startIndex);
           return;
@@ -204,7 +204,7 @@ export const useFieldV7TextField = params => {
       return;
     }
     setFocused(true);
-    const isFocusInsideASection = sectionListRef.current.getSectionIndexFromDOMElement(getActiveElement(document)) != null;
+    const isFocusInsideASection = sectionListRef.current.getSectionIndexFromDOMElement(getActiveElement(sectionListRef.current?.getRoot?.()?.ownerDocument)) != null;
     if (!isFocusInsideASection) {
       setSelectedSections(sectionOrder.startIndex);
     }
@@ -215,7 +215,7 @@ export const useFieldV7TextField = params => {
       if (!sectionListRef.current) {
         return;
       }
-      const activeElement = getActiveElement(document);
+      const activeElement = getActiveElement(sectionListRef.current?.getRoot?.()?.ownerDocument);
       const shouldBlur = !sectionListRef.current.getRoot().contains(activeElement);
       if (shouldBlur) {
         setFocused(false);
